File: leetcode-go/78.子集.go
/*
 * @lc app=leetcode.cn id=78 lang=golang
 *
 * [78] 子集
 */
package main

import "fmt"

func main() {
	nums := []int{1, 2, 3}
	res := subsets(nums)
	fmt.Println(res)
}

// @lc code=start
func subsets(nums []int) [][]int {
	// ans := make([][]int, 0)
	// n := len(nums)
	// for mask := 0; mask < 1<<n; mask++ {
	// 	set := []int{}
	// 	for i, v := range nums {
	// 		// mask 第i位是否为1
	// 		if mask>>i&1 > 0 {
	// 			set = append(set, v)
	// 		}
	// 	}
	// 	ans = append(ans, append([]int(nil), set...))
	// }
	// return ans

	ans := make([][]int, 1<<len(nums))
	for i := range ans { // 枚举全集 U 的所有子集 i
		for j, x := range nums {
			if i>>j&1 == 1 { // j 在集合 i 中
				ans[i] = append(ans[i], x)
			}
		}
	}

	return ans
}

// @lc code=end

File: 78.子集.py
#
# @lc app=leetcode.cn id=78 lang=python3
#
# [78] 子集
#
from typing import List


# @lc code=start
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        # res = []
        # n = len(nums)

        # for i in range(1 << n):
        #     tmp = []
        #     for j in range(n):
        #         if ((i >> j) & 1) > 0:
        #             tmp.append(nums[j])
        #     res.append(tmp)
        # return res
        res = [[]]
        for i in range(len(nums)):
            for j in range(len(res)):
                res.append(res[j]+[nums[i]])
        return res
# @lc code=end
nums = [1, 2, 3]
Solution().subsets(nums)
